da/celestia/celestia.go

```diff
@@ -264,71 +265,103 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
 
 func (c *DataAvailabilityLayerClient) RetrieveBatches(daMetaData *da.DASubmitMetaData) da.ResultRetrieveBatch {
 
-	//Just for backward compatibility, in case no commitments are sent from the Hub, batch can be retrieved using previous implementation.
-	if daMetaData.Commitment == nil {
-		return c.retrieveBatches(daMetaData.Height)
-	}
-
 	for {
 		select {
 		case <-c.ctx.Done():
 			c.logger.Debug("Context cancelled")
 			return da.ResultRetrieveBatch{}
 		default:
-			var batches []*types.Batch
-			//for _, commitment := range daMetaData.Commitments {
-			blob, err := c.rpc.Get(c.ctx, daMetaData.Height, c.config.NamespaceID.Bytes(), daMetaData.Commitment)
-			if err != nil {
-				return da.ResultRetrieveBatch{
-					BaseResult: da.BaseResult{
-						Code:    da.StatusError,
-						Message: err.Error(),
-						Error:   da.ErrBlobNotFound,
-					},
-				}
-			}
-			if blob == nil {
-				return da.ResultRetrieveBatch{
-					BaseResult: da.BaseResult{
-						Code:    da.StatusError,
-						Message: "Blob not found",
-						Error:   da.ErrBlobNotFound,
-					},
+			//Just for backward compatibility, in case no commitments are sent from the Hub, batch can be retrieved using previous implementation.
+			var resultRetrieveBatch da.ResultRetrieveBatch
+			err := retry.Do(func() error {
+				var result da.ResultRetrieveBatch
+				if daMetaData.Commitment == nil {
+					result = c.retrieveBatchesNoCommitment(daMetaData.Height)
+				} else {
+					result = c.retrieveBatches(daMetaData)
 				}
-			}
+				resultRetrieveBatch = result
 
-			var batch pb.Batch
-			err = proto.Unmarshal(blob.Data, &batch)
-			if err != nil {
-				c.logger.Error("failed to unmarshal block", "daHeight", daMetaData.Height, "error", err)
-			}
-			parsedBatch := new(types.Batch)
-			err = parsedBatch.FromProto(&batch)
-			if err != nil {
-				return da.ResultRetrieveBatch{
-					BaseResult: da.BaseResult{
-						Code:    da.StatusError,
-						Message: err.Error(),
-						Error:   err,
-					},
+				if result.Error == da.ErrRetrieval {
+					c.logger.Error("Failed in retrieving blob")
+					return result.Error
 				}
-			}
-			batches = append(batches, parsedBatch)
-			//}
-			return da.ResultRetrieveBatch{
-				BaseResult: da.BaseResult{
-					Code:    da.StatusSuccess,
-					Message: "Batch retrieval successful",
-				},
-				Batches: batches,
-			}
+
+				return nil
+			}, retry.Attempts(uint(c.rpcRetryAttempts)), retry.DelayType(retry.FixedDelay), retry.Delay(c.rpcRetryDelay))
+			c.logger.Error("RetrieveBatches process failed", "error", err)
+
+			return resultRetrieveBatch
+
+		}
+	}
+
+}
+
+func (c *DataAvailabilityLayerClient) retrieveBatches(daMetaData *da.DASubmitMetaData) da.ResultRetrieveBatch {
+
+	ctx, cancel := context.WithTimeout(c.ctx, c.config.Timeout)
+	defer cancel()
```

Comment:
the `defer cancel()` is redundent all over the code, as it all runs in same goroutine in sync order

Reply:
i added the defer cancel(), otherwise it raises the following lint warning: lostcancel: the cancel function returned by context.WithTimeout should be called, not discarded, to avoid a context leak (govet). Not sure what is the best practice to avoid that without being redundant.

Reply:
@mtsitrin basically it cancels the local context created if there was no timeout (so the context won't persist after successful execution). why is it redundant?

Reply:
by bad, confused with non timeout context. pardon